text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+The FPC Shim enables an FPC Chaincode to talk to the Peer, in particular to invoke the chaincode and provide "normal" shim operations such as `getState` and `putState`.
+The FPC SDK provides the FPC Shim to enable FPC chaincode developers to easily link to this library when compiling their FPC Chaincode and building a deployment artifact. It implements the interface from the Peer to the FPC Chaincode, encrypts and decrypts state and arguments, and implements Attestation.
+See the [FPC Shim](#fpc-shim) section for more information.
+
+**Ledger Enclave:** The Ledger Enclave is a crucial component to protect the execution of a FPC chaincode.
+In particular, it is responsible to ensure that a FPC chaincode processes committed world state data, i.e. only state data that comes from valid and committed transactions.
+The Ledger Enclave is a separate enclave that establishes a trusted view of the ledger and locally maintains integrity metadata for validating information on the blockchain ledger.
+Like the Peer, it performs standard validation logic when a new block of transactions arrives, but it also creates and stores a cryptographic hash of each new key-value pair of the blockchain state. This makes it possible to verify that the data coming from the Peer is correct -- remember that in the FPC setting we do not necessarily trust a (single) peer. The Chaincode Enclave uses the Ledger Enclave as its tamperproof source of information on the blockchain world state. The Ledger Enclave and a Chaincode Enclave interact by through a secure channel that is initially established when the Chaincode Enclave is deployed. There is one Ledger Enclave per Channel on any given Peer. When a peer receives a new block from the ordering service, it forwards it to the Ledger Enclave. The Ledger Enclave is part of the FPC runtime.
+
+**FPC Registry:** Also referred to as the Enclave Registry Chaincode (ERCC), this is a component which maintains a list of all Chaincode Enclaves deployed on the peers in a channel.
+The registry associates with each enclave their identity, associated public keys and an attestation linking them. Additionally, the registry manages chaincode specific keys, including a chaincode public encryption key, and facilitates corresponding key-management among authorized Chaincode Enclaves. Lastly, the registry also records information required to bootstrap the validation of attestation. All of this information is committed firmly on the ledger. This enables any Peer in the Channel (even those without SGX) to inspect the Attestation results before taking actions such as connecting to that chaincode or committing transactions produced by an FPC chaincode. Moreover, clients can query the registry to retrieve the chaincode public encryption keys of a particular FPC chaincode so they can send privately transaction proposals for endorsement.
```

Comment:
> Moreover, clients can query the registry to retrieve the chaincode public encryption keys of a particular FPC chaincode so they can send privately transaction proposals for endorsement.

But you would need the clients to have signature public keys of the enclaves, otherwise the peer could do a MITM, no? 

So, why not assume clients have the encryption public keys in the first place? Or, do you assume there is some kind of PKI in place that certifies the signature public keys of the enclaves?

Reply:
Note we assume the standard fabric model that clients talk to peer of their own organization for queries. So no MITM in that case and also no need to get the signature keys of enclaves.  If you would extend the model to clients outside of peer orgs, then it gets a bit more complicated but this would be true also for vanilla fabric (i.e., each query would have be replicated consistent with the implied endorsement policy. )

Also note the SGX attestation authority essentially roots a PKI, although, for efficiency and simplicity the Enclave Registry forms its own (sub)PKI but could always be verified also by clients to be grounded in SGX attestation and appropriate measurements .

Reply:
> Note we assume the standard fabric model that clients talk to peer of their own organization for queries. So no MITM in that case and also no need to get the signature keys of enclaves.

So we trust the a peer won't be malicious against a client from its own organization?

Reply:
That's correct.  As mentioned, it is not essential and the architecure could be extended to clients outside of orgs but the initial approach should suffice for what we would foresee the most common scenarios in a permissioned ledger.